bin/describe.ml

```diff
@@ -747,61 +752,133 @@ module Preprocess = struct
     in
     let pp_file =
-      file_in_build_dir |> Path.map_extension ~f:(fun ext -> ".pp" ^ ext)
+      file_in_build_dir |> Path.Build.map_extension ~f:(fun ext -> ".pp" ^ ext)
     in
-    Build_system.file_exists pp_file >>= function
+    let* project = Source_tree.root () >>| Source_tree.Dir.project in
+    let* pp_file_exists = Build_system.file_exists (Path.build pp_file) in
+    match pp_file_exists with
     | true ->
-      let* () = Build_system.build_file pp_file in
-      let+ project = Source_tree.root () >>| Source_tree.Dir.project in
-      Ok (project, pp_file)
+      let pp_file = Path.build pp_file in
+      let+ () = Build_system.build_file pp_file in
+      Ok (`Pp_file (project, pp_file))
     | false -> (
-      Build_system.file_exists file_in_build_dir >>= function
+      let* file_exists =
+        Build_system.file_exists (Path.build file_in_build_dir)
+      in
+      match file_exists with
+      | false ->
+        User_error.raise
+          [ Pp.textf "%s does not exist"
+              (Path.Build.to_string file_in_build_dir)
+          ]
       | true -> (
-        let* dir =
-          Source_tree.nearest_dir (Path.Source.of_string file)
-          >>| Source_tree.Dir.path >>| Path.source
+        let module_name =
+          let open Option.O in
+          let* name =
+            let fname = Path.Build.basename file_in_build_dir in
+            let name = Filename.remove_extension fname in
+            if String.equal fname name then None else Some name
+          in
+          External_lib_deps.Module_name.of_string_opt name
         in
-        let* dune_file =
-          External_lib_deps.Dune_load.Dune_files.in_dir (dir |> in_build_dir)
+        let* stanza =
+          let drop_rules f =
+            let+ res, _ =
+              Memo.Implicit_output.collect Dune_engine.Rules.implicit_output f
+            in
+            res
+          in
+          match module_name with
+          | None -> Memo.return None
+          | Some module_name ->
+            Path.Build.parent file_in_build_dir
+            |> Memo.Option.bind ~f:(fun dir ->
+                   let* dir_contents =
+                     drop_rules @@ fun () ->
+                     External_lib_deps.Dir_contents.get super_context ~dir
+                   in
+                   let* ocaml =
+                     External_lib_deps.Dir_contents.ocaml dir_contents
+                   in
+                   let stanza =
+                     match
+                       External_lib_deps.Ml_sources.find_origin ocaml
+                         module_name
+                     with
+                     | Some
+                         (External_lib_deps.Ml_sources.Origin.Executables exes)
+                       -> Some (`Executables exes)
+                     | Some (Library lib) -> Some (`Library lib)
+                     | None | Some (Melange _) -> None
+                   in
+                   Memo.return stanza)
```

Comment:
This is extracted from https://github.com/ocaml/dune/blob/main/src/dune_rules/top_module.ml#L16. Should we try to extract the common part into its own function?

Reply:
I think you can just use it directly already. You can then use `Compilation_context.preprocessing` to get the preprocessing spec.

Reply:
I extracted the common part in [`a4972fa` (#7522)](https://github.com/ocaml/dune/pull/7522/commits/a4972fa59a7224e5139030abd9ec589553f59531). I'm not fan of the implementation, though (see inline comments).